File: compromised.py
"""Track compromised recovery phrases by hash only — never store raw seeds."""

from __future__ import annotations

import hashlib
import json
from pathlib import Path

STORE_PATH = Path(__file__).resolve().parent / "data" / "compromised_hashes.json"


def phrase_hash(phrase: str) -> str:
    normalized = " ".join(phrase.lower().split())
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()


def _load() -> set[str]:
    if not STORE_PATH.exists():
        return set()
    try:
        data = json.loads(STORE_PATH.read_text(encoding="utf-8"))
        return set(data.get("hashes", []))
    except (OSError, json.JSONDecodeError, TypeError):
        return set()


def _save(hashes: set[str]) -> None:
    STORE_PATH.parent.mkdir(parents=True, exist_ok=True)
    STORE_PATH.write_text(
        json.dumps({"hashes": sorted(hashes)}, indent=2),
        encoding="utf-8",
    )


def mark_compromised(phrase: str) -> str:
    """Label a phrase as compromised. Stores SHA-256 only."""
    digest = phrase_hash(phrase)
    hashes = _load()
    hashes.add(digest)
    _save(hashes)
    return digest


def is_compromised(phrase: str) -> bool:
    return phrase_hash(phrase) in _load()

File: lib/compromised.ts
"use client";

/** SHA-256 hashes of compromised phrases — never store raw seeds. */

const KEY = "keystone-compromised-hashes";

async function sha256(text: string): Promise<string> {
  const data = new TextEncoder().encode(text);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return [...new Uint8Array(digest)].map((b) => b.toString(16).padStart(2, "0")).join("");
}

function load(): Set<string> {
  try {
    const raw = localStorage.getItem(KEY);
    if (!raw) return new Set();
    return new Set(JSON.parse(raw) as string[]);
  } catch {
    return new Set();
  }
}

function save(hashes: Set<string>) {
  localStorage.setItem(KEY, JSON.stringify([...hashes]));
}

export async function phraseHash(phrase: string): Promise<string> {
  const normalized = phrase.toLowerCase().trim().split(/\s+/).filter(Boolean).join(" ");
  return sha256(normalized);
}

export async function markCompromised(phrase: string): Promise<string> {
  const digest = await phraseHash(phrase);
  const hashes = load();
  hashes.add(digest);
  save(hashes);
  return digest;
}

export async function isCompromised(phrase: string): Promise<boolean> {
  const digest = await phraseHash(phrase);
  return load().has(digest);
}
